Tv casting app stop connect when user navigates back before confirming passcode in commission

yuenton-amazon opened the pull request:
## Updated Android tv-casting-app to send stopConnect command when user exits UDC before confirming passcode

### Changes
1. Added IsPendingPasscodeFromUser() to CastingPlayer. This function
   will return true if we are still connecting and pending user action
   for passcode
2. ConnectionExampleFragment.java onDestroy() calls stopConnect() if user
   has not confirmed passcode
4. MCConnectionExampleView ConnectingView dissapear calls stopConnect()
   if user has not confirmed passcode

### Testing
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Updated Android & iOS tv-casting-app to send stopConnect command when user
exits UDC before confirming passcode

Changes
1. Added IsPendingPasscodeFromUser() to CastingPlayer. This function
   will return true if we are still connecting and pending user action
   for passcode
2. ConnectionExampleFragment.java onDestroy() calls stopConnect() if user
   has not confirmed passcode
4. MCConnectionExampleView ConnectingView dissapear calls stopConnect()
   if user has not confirmed passcode

Test
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37278) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37278) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=37278) it.</sub>

sharadb-amazon commented on examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp:
```diff
@@ -228,11 +228,6 @@ CHIP_ERROR CastingPlayer::ContinueConnecting()
 
 CHIP_ERROR CastingPlayer::StopConnecting()
 {
-    VerifyOrReturnValue(
-        mIdOptions.mCommissionerPasscode, CHIP_ERROR_INCORRECT_STATE,
-        ChipLogError(AppServer,
-                     "CastingPlayer::StopConnecting() mIdOptions.mCommissionerPasscode == false, ContinueConnecting() should only "
-                     "be called when the CastingPlayer/Commissioner-Generated passcode commissioning flow is in progress."););
```
@pgregorr-amazon - Correct me if I am wrong, but I remember you added this check to guard against some error. Is it safe to remove this?

pgregorr-amazon replied:
@sharadb-amazon – This check wasn’t guarding against a specific error but was added as a defensive measure to restrict `CastingPlayer::StopConnecting()` API to the CommissioneR-Generated passcode flow. The CommissioneE-Generated passcode flow wasn’t tested at the time, and there was no use case for it. Before removing this check, we should verify that the API functions correctly with the CommissioneE-Generated passcode flow.
cc @yuenton-amazon

sharadb-amazon replied:
@yuenton-amazon  can you please verify the above?

yuenton-amazon replied:
Talked to Philip. He mentioned he checks that the tv-app's log to check if its received the udc cancel message. I've already verified that in commissionee generated passcode, stop connect results in tv-app receiving the message and removes the udc session with `controller udc-print` CLI command

sharadb-amazon commented on examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/ConnectionExampleFragment.java:
```diff
@@ -42,6 +42,14 @@
 
 /** A {@link Fragment} to Verify or establish a connection with a selected Casting Player. */
 public class ConnectionExampleFragment extends Fragment {
+  enum ConnectionState {
+      PENDING_PASSCODE_CONFIRMATION,
+      CONNECTING,
+      CONNECTED,
+      FAILED,
+      CANCELLED,
+  }
```
Should this be added in the APIs themselves? Like for Android, here: https://github.com/project-chip/connectedhomeip/blob/master/examples/tv-casting-app/android/App/app/src/main/jni/com/matter/casting/core/CastingPlayer.java

Note that we have a similar C++ enum in the tv-casting-common layer: https://github.com/project-chip/connectedhomeip/blob/master/examples/tv-casting-app/tv-casting-common/core/CastingPlayer.h#L81-L86

yuenton-amazon replied:
The important state to trigger `stopConnecting()` is actually only `PENDING_PASSCODE_CONFIRMATION`. Since after after `continueConnect()` is called in `CONNECTING` state, `stopConnecting()` cannot be called.

The condition to call  stopConnecting() is `PENDING_PASSCODE_CONFIRMATION`. I can covert this to a bool to simplify the solution, like the approach taken on iOS side.

yuenton-amazon replied:
Will update to use CastingPlayer.h states for both iOS and Android

sharadb-amazon reviewed (commented):
Can you fix these CI check errors:
1. Restyler: https://github.com/project-chip/connectedhomeip/actions/runs/13037028900/job/36370278499?pr=37278
2. https://github.com/project-chip/connectedhomeip/actions/runs/13037028875/job/36370271674?pr=37278

pgregorr-amazon commented on examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/ConnectionExampleFragment.java:
```diff
@@ -262,6 +262,22 @@ public void handle(CommissionerDeclaration cd) {
+    super.onDestroy();
+
+    // Only stop connection if we are pending passcode confirmation
+    // We cannot stop connection once continueConnecting() is called
```
We should be able to safely call stopConnecting() anytime.

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Updated native layer and iOS tv-casting-app to use native layer
ConnectionState instead of a pendingPasscode customer check.

Changes
* Expose native layer ConnectionState to iOS tv-casting-app
* Update iOS tv-casting-app to use ConnectionState to determine if we
  should call stopConnect()
* Only call stopConnect() on navigate back from ConnectionExampleView
* Updated doc to indicate when a object method will free itself as a
  side effect

Test
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Updated JNI and Android tv-casting-app to use native layer
ConnectionState instead of a pendingPasscode customer check.

Changes
* Expose native layer ConnectionState to Android tv-casting-app
* Update Android tv-casting-app to use ConnectionState to determine if we
  should call stopConnect()
* Only call stopConnect() on navigate back from
  ConnectionExampleFragmnet

Test
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Applied restyle formatting

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Fix restyle CI error

yunhanw-google commented on examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCastingPlayer-JNI.cpp:
```diff
@@ -165,6 +165,32 @@ JNI_METHOD(void, disconnect)
+    switch (state)
+    {
+    case matter::casting::core::ConnectionState::CASTING_PLAYER_NOT_CONNECTED:
+        return env->NewStringUTF("NOT_CONNECTED");
```
please try to use CharToStringUTF, context is from
https://github.com/project-chip/connectedhomeip/blob/1612d529cd3cb4e1416f36d4dcf47c9ab57873cd/src/lib/support/JniReferences.h#L237

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Updated JNI and Android tv-casting-app to use CharToStringUTF
util function to create jstring and handle NULL JNIEnv interface

Changes
* JNI to use CharToStringUTF util function to create jstring
* Updated documentation for new JNI functions

Test
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

yuenton-amazon committed:
[Amazon] StopConnect for UDC in Android tv-casting-app

Updated JNI and Android tv-casting-app to use CharToStringUTF
util function to create jstring and handle NULL JNIEnv interface

Changes
* JNI to use CharToStringUTF util function to create jstring for all
  ConnectionState and not just errors

Test
1. Manually verified UDC attempt is successful both Android iOS for following
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Navigate back to discovery page
   3. Start new UDC attempt (both commissionee & commissioner generated passcode)

2. Manually verified UDC attempt is successful both Android iOS for following. UDC will finish in the background.
   1. Attempt UDC attempt (both commissionee & commissioner generated passcode)
   2. Confirm passcode on TV app or casting app depending on who
      generated passcode
   3. Navigate back before commission fnishes.

3. Manual regression test following
   * UDC attempt happy path (no navigate back) for both commissinee &
     commissioner generated passcode

yuenton-amazon commented on examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCastingPlayer-JNI.cpp:
```diff
@@ -165,6 +165,32 @@ JNI_METHOD(void, disconnect)
+JNI_METHOD(jstring, getConnectionStateNative)
+(JNIEnv * env, jobject thiz)
+{
+    chip::DeviceLayer::StackLock lock;
```
Done
